In [1]:
import pandas as pd
df = pd.read_csv('spaceship_tt.csv')

In [2]:
df['CryoSleep'] = df['CryoSleep']*1
df['Transported'] = df['Transported']*1
df['VIP'] = df['VIP']*1
df.isna().sum()


PassengerId       0
HomePlanet      201
CryoSleep       217
Cabin           199
Destination     182
Age             179
VIP             203
RoomService     181
FoodCourt       183
ShoppingMall    208
Spa             183
VRDeck          188
Name            200
Transported       0
dtype: int64

- 결측치를 가장 나이스하게 전처리했는가가 성능에 중요한 기준이 된다.
- 모델링 했는데 성능이 안 나왔다는 것은 전처리를 다시 확인해봐야한다.

# 전처리X 베이스 모델로 정확도 평가

In [3]:
df_base=df[['CryoSleep','Age','VIP','RoomService','FoodCourt','ShoppingMall','Spa','VRDeck','Transported']]
df_base.dropna(inplace=True)
df_base

<ipython-input-3-1256e4537ec4>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_base.dropna(inplace=True)


CryoSleep   Age VIP  RoomService  FoodCourt  ShoppingMall     Spa  \
0            0  39.0   0          0.0        0.0           0.0     0.0   
1            0  24.0   0        109.0        9.0          25.0   549.0   
2            0  58.0   1         43.0     3576.0           0.0  6715.0   
3            0  33.0   0          0.0     1283.0         371.0  3329.0   
4            0  16.0   0        303.0       70.0         151.0   565.0   
...        ...   ...  ..          ...        ...           ...     ...   
8688         0  41.0   1          0.0     6819.0           0.0  1643.0   
8689         1  18.0   0          0.0        0.0           0.0     0.0   
8690         0  26.0   0          0.0        0.0        1872.0     1.0   
8691         0  32.0   0          0.0     1049.0           0.0   353.0   
8692         0  44.0   0        126.0     4688.0           0.0     0.0   

      VRDeck  Transported  
0        0.0            0  
1       44.0            1  
2       49.0            0  
3      193.0            0  
4        2.0            1  
...      ...          ...  
8688    74.0            0  
8689     0.0            0  
8690     0.0            1  
8691  3235.0            0  
8692    12.0            1  

[7251 rows x 9 columns]

In [5]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
dt_clf = DecisionTreeClassifier(random_state=111)
X_train, X_test, y_train, y_test = train_test_split(df_base.drop('Transported',axis=1),df_base['Transported'],
                                                   test_size=0.3, random_state=111)
dt_clf = dt_clf.fit(X_train,y_train)

In [6]:
X_train

CryoSleep   Age VIP  RoomService  FoodCourt  ShoppingMall    Spa  VRDeck
5664         0  10.0   0          0.0        0.0           0.0    0.0     0.0
6303         0  28.0   0         32.0        0.0         304.0  291.0     2.0
963          0   0.0   0          0.0        0.0           0.0    0.0     0.0
6375         0  21.0   0       1244.0        0.0        1119.0    0.0     3.0
8299         0  34.0   0          0.0     3126.0           0.0  112.0    15.0
...        ...   ...  ..          ...        ...           ...    ...     ...
5017         1  38.0   0          0.0        0.0           0.0    0.0     0.0
5787         0  33.0   1          0.0     3095.0           0.0  197.0    40.0
2429         1  34.0   0          0.0        0.0           0.0    0.0     0.0
8421         0  39.0   0       2101.0       41.0           0.0  525.0    19.0
4717         1  13.0   0          0.0        0.0           0.0    0.0     0.0

[5075 rows x 8 columns]

In [7]:
#실제 예측값 얼마인지?
from sklearn.metrics import accuracy_score
#하이퍼파라미터 튜닝하지 않은 base_model
pred = dt_clf.predict(X_test)
accuracy = accuracy_score(y_test, pred)
print(accuracy)
print(dt_clf.get_params())

0.7601102941176471
{'ccp_alpha': 0.0, 'class_weight': None, 'criterion': 'gini', 'max_depth': None, 'max_features': None, 'max_leaf_nodes': None, 'min_impurity_decrease': 0.0, 'min_samples_leaf': 1, 'min_samples_split': 2, 'min_weight_fraction_leaf': 0.0, 'random_state': 111, 'splitter': 'best'}


# My csv 파일로 모델링 진행

In [15]:
mine = pd.read_csv('/content/문다은87_0730_과제2파일.csv')

In [16]:
mine['CryoSleep'] =mine['CryoSleep']*1
mine['VIP'] =mine['VIP']*1

In [18]:
mine['Transported'] =mine['Transported']*1

In [20]:
minedt_clf1 = DecisionTreeClassifier(random_state=111)

X_train, X_test, y_train, y_test = train_test_split(mine.drop('Transported',axis=1),mine['Transported'],
                                                   test_size=0.3, random_state=111)

minedt_clf1 = minedt_clf1.fit(X_train,y_train)

#실제 예측값 얼마인지?
from sklearn.metrics import accuracy_score
#하이퍼파라미터 튜닝하지 않은 base_model
pred1 = minedt_clf1.predict(X_test)
accuracy = accuracy_score(y_test, pred1)
print(accuracy)
print(minedt_clf1.get_params())

0.7588134135855547
{'ccp_alpha': 0.0, 'class_weight': None, 'criterion': 'gini', 'max_depth': None, 'max_features': None, 'max_leaf_nodes': None, 'min_impurity_decrease': 0.0, 'min_samples_leaf': 1, 'min_samples_split': 2, 'min_weight_fraction_leaf': 0.0, 'random_state': 111, 'splitter': 'best'}


- 전처리를 안하고도 0.7601102941176471
- 전처리를 했는데 0.7588134135855547 정확도 나온다.
- 성능이 높아지지 않았다!!
- 타겟과의 관계를 고려하고, 실제 데이터들의 시각화를 통해 분포 및 범주들을 확인하고 변수를 추가해야 한다.
- 많은 변수들을 추가한다는게 더 높은 성능을 보이는 것은 아니다.
- 결측치를 더 잘 처리하는 것도 정말 중요하다


# Plus Idea

In [33]:
df['psid_cnt']= df['PassengerId'].str[:4]

In [34]:
#전처리를 원하는 건 한 공간에 몇 명이 같이 탑승했는가?
df_psid=pd.DataFrame(df.groupby('psid_cnt')['PassengerId'].count()).reset_index()

In [35]:
df_psid

psid_cnt  PassengerId
0        0001            1
1        0002            1
2        0003            2
3        0004            1
4        0005            1
...       ...          ...
6212     9275            3
6213     9276            1
6214     9278            1
6215     9279            1
6216     9280            2

[6217 rows x 2 columns]

In [36]:
df_pre1 =pd.merge(df, df_psid, left_on= 'psid_cnt',right_on='psid_cnt', how='inner')

In [37]:
df_pre1.PassengerId_y.value_counts()

1    4805
2    1682
3    1020
4     412
5     265
7     231
6     174
8     104
Name: PassengerId_y, dtype: int64

In [38]:
df_pre1['PassengerId_y'].isna().sum()

0

In [39]:
df_pre1.columns

Index(['PassengerId_x', 'HomePlanet', 'CryoSleep', 'Cabin', 'Destination',
       'Age', 'VIP', 'RoomService', 'FoodCourt', 'ShoppingMall', 'Spa',
       'VRDeck', 'Name', 'Transported', 'psid_cnt', 'PassengerId_y'],
      dtype='object')

In [40]:
df_pre1=df_pre1[['HomePlanet', 'CryoSleep', 'Cabin', 'Destination',
       'Age', 'VIP', 'RoomService', 'FoodCourt', 'ShoppingMall', 'Spa',
       'VRDeck', 'Name', 'Transported', 'PassengerId_y']]

In [41]:
df_pre1=df_pre1[['CryoSleep','Age','VIP','RoomService','FoodCourt','ShoppingMall','Spa','VRDeck','Transported','PassengerId_y']]

In [30]:
df_pre1

CryoSleep   Age VIP  RoomService  FoodCourt  ShoppingMall     Spa  \
0            0  39.0   0          0.0        0.0           0.0     0.0   
1            0  24.0   0        109.0        9.0          25.0   549.0   
2            0  58.0   1         43.0     3576.0           0.0  6715.0   
3            0  33.0   0          0.0     1283.0         371.0  3329.0   
4            0  16.0   0        303.0       70.0         151.0   565.0   
...        ...   ...  ..          ...        ...           ...     ...   
8688         0  41.0   1          0.0     6819.0           0.0  1643.0   
8689         1  18.0   0          0.0        0.0           0.0     0.0   
8690         0  26.0   0          0.0        0.0        1872.0     1.0   
8691         0  32.0   0          0.0     1049.0           0.0   353.0   
8692         0  44.0   0        126.0     4688.0           0.0     0.0   

      VRDeck  Transported  PassengerId_y  
0        0.0            0              1  
1       44.0            1              1  
2       49.0            0              2  
3      193.0            0              2  
4        2.0            1              1  
...      ...          ...            ...  
8688    74.0            0              1  
8689     0.0            0              1  
8690     0.0            1              1  
8691  3235.0            0              2  
8692    12.0            1              2  

[8693 rows x 10 columns]

In [42]:
df_pre1.dropna(inplace=True)


In [43]:
dt_clf6 = DecisionTreeClassifier(random_state=111)

X_train, X_test, y_train, y_test = train_test_split(df_pre1.drop('Transported',axis=1),df_pre1['Transported'],
                                                   test_size=0.3, random_state=111)

dt_clf6 = dt_clf6.fit(X_train,y_train)

#실제 예측값 얼마인지?
from sklearn.metrics import accuracy_score
#하이퍼파라미터 튜닝하지 않은 base_model
pred6 = dt_clf6.predict(X_test)
accuracy = accuracy_score(y_test, pred6)
print(accuracy)
print(dt_clf6.get_params())

0.7463235294117647
{'ccp_alpha': 0.0, 'class_weight': None, 'criterion': 'gini', 'max_depth': None, 'max_features': None, 'max_leaf_nodes': None, 'min_impurity_decrease': 0.0, 'min_samples_leaf': 1, 'min_samples_split': 2, 'min_weight_fraction_leaf': 0.0, 'random_state': 111, 'splitter': 'best'}


In [44]:
# 방금 새로운 컬럼 추가 0.7463235294117647
# 베이스라인 0.7601102941176471

# 파생변수 하나 추가했는데 오히려 더 떨어진 거 아닌가? 정확도가 더 떨어졌다.
# 왜 떨어졌나?
pd.DataFrame(df_pre1.groupby(['PassengerId_y','Transported'])['VIP'].count()).reset_index()

PassengerId_y  Transported   VIP
0               1            0  2198
1               1            1  1813
2               2            0   652
3               2            1   746
4               3            0   355
5               3            1   501
6               4            0   119
7               4            1   223
8               5            0    94
9               5            1   125
10              6            0    54
11              6            1    94
12              7            0    82
13              7            1   109
14              8            0    54
15              8            1    32

- 파생변수를 만들거나 전처리를 진행할 때
- 모든 변수를 다 사용할 필요도 없고, 파생변수를 만들 때도 정답 레이블과 비교하여 정답레이블을 잘 나눌 수 있는 이 부분을 꼭 잘 이해해야 한다.